### ISSUE TITLE
`dependencyTree` should also display/write json, not just text

### ISSUE THREAD

**Issue by mkurz** [sbt/sbt#7770]
`dependencyTree` currently only displays text, but it would be nice to also be able to output json.
Same with writing to file, I would like to write a json file: `dependencyTree/toFile /some/folder/deps.json`

The funny thing is `dependencyBrowseTreeHTML` already creates a `target/tree.json` which is used as source to display the dependencies in the html/browser. With `jq` you can format it nicely as well: `jq '.' target/tree.json`

So in theory as a workaround I can already do what this issue asks to with running `dependencyBrowseTreeHTML` but it's not obivous.

Making it more explicit with the `dependencyTree` command as a flag would be nicer IMHO.

Related:

- #7771

**Comment by mzuehlke** [comment]
The "original" `addDependencyTreePlugin ` https://github.com/sbt/sbt-dependency-graph#main-tasks had a `toFile ` subtask. Is this still functional ?

**Comment by mkurz** [comment]
@mzuehlke  yes, it still works. I also mentioned it in my initial post. However it does not write json.

**Comment by mzuehlke** [comment]
sorry @mkurz , I should have read your post more carefully 🙈

**Comment by eed3si9n** [comment]
Since `dependencyTree` is a task targeting humans (as opposed to other downstream tasks), I wonder if we should make it into an input task and let it take various command line parameters like `--format json`. That can also reduce the number of tasks.

**Comment by Friendseeker** [comment]
What type of args should `dependencyTree` accept?

I am thinking of the following

```
--format (accepts text/json, decides format of output, default to text format if not supplied)
--writeto (accepts absolute path / relative path to a directory, writes text/json output to the path)
```

**Comment by mkurz** [comment]
@Friendseeker Not sure if you are aware but you can already call
```sbt
dependencyTree/toFile /some/folder/output.txt
```
Also you can override the file with `-f` (force):
```sbt
dependencyTree/toFile /some/folder/output.txt -f
```

**Comment by Friendseeker** [comment]
@mkurz @eed3si9n I did a draft Input Task based implementation at https://github.com/Friendseeker/sbt-dependency-report, with the aims of

- Only featuring 1 single input task (avoiding settings key explosion)
- Support all capabilities of `MiniDependencyTreePlugin`
- Support most capabilities of the full `DependencyTreePlugin`
- Support json output to both terminal & file
- has a very simple API

Currently its API is the following

> --format=\<format\>: Specifies the format of the report. Supported formats are:
> text: Default format with options like list, stats, and info.
> json: Outputs the report in JSON format.
> html: Outputs the report in HTML format.
> graphml: Outputs the report in GraphML format.
> --toFile=\<path\>: Specifies the file path for saving the report. If not provided, the report is saved in the target directory.
> --config=\<scope\>: Specifies the scope of the dependency graph to generate, either Compile or Test.

For example

```shell
sbt:contraband> dependencyReport
[info] org.scala-sbt:contraband_2.12:0.5.3-SNAPSHOT [S]
[info]   +-com.eed3si9n:sjson-new-scalajson_2.12:0.10.0 [S]
[info]   | +-com.eed3si9n:shaded-jawn-parser_2.12:1.3.2 [S]
[info]   | +-com.eed3si9n:shaded-scalajson_2.12:1.0.0-M4 [S]
[info]   | +-com.eed3si9n:sjson-new-core_2.12:0.10.0 [S]
[info]   |
[info]   +-org.parboiled:parboiled_2.12:2.5.1 [S]
[info]
[info] Dependency report written to D:\Repos\contraband\library\target\dependencies.txt

sbt:contraband> dependencyReport --format=html
[info] Dependency report written to D:\Repos\contraband\library\target\dependencies.html

sbt:contraband> dependencyReport --format=json --toFile=output.json
[info] [{"text":"org.scala-sbt:contraband_2.12:0.5.3-SNAPSHOT","children":[{"text":"com.eed3si9n:sjson-new-scalajson_2.12:0.10.0","children"
:[{"text":"com.eed3si9n:shaded-jawn-parser_2.12:1.3.2","children":[]},{"text":"com.eed3si9n:shaded-scalajson_2.12:1.0.0-M4","children":[]},{"text":"com.eed3si9n:sjson-new-core_2.12:0.10.0","children":[]}]},{"text":"org.parboiled:parboiled_2.12:2.5.1","children":[]}]}]
[info] Dependency report written to D:\Repos\contraband\output.json
```

I want to ask for some early feedbacks around the UX, before consolidating the implementation & in-sourcing it to sbt repo.

To try it out locally it suffices to clone the repo, run `publishLocal` then add `addSbtPlugin("com.Friendseeker" % "sbt-dependency-report" % "0.1.0-SNAPSHOT")` to an sbt project.

**Comment by mkurz** [comment]
I am wondering if the `--toFile=` is still needed if you now by default always write the file in the target folder anyway.

Reason is because if you have multi-project builds, currently, with sbt 1.10.x, (when using `toFile`) it will just write the file for the first project and then complain for the next projects. Example from Play Framework:

```sbt
sbt:Play-Framework> dependencyTree/toFile /tmp/deps.txt
...
[error] (Play-JDBC / Compile / dependencyTree / toFile) Target file for dependencyTree already exists at /tmp/deps.txt. Use '-f' to override
[error] (Play-Integration-Test / Compile / dependencyTree / toFile) Target file for dependencyTree already exists at /tmp/deps.txt. Use '-f' to override
[error] (Play-Pekko-Http-Server / Compile / dependencyTree / toFile) Target file for dependencyTree already exists at /tmp/deps.txt. Use '-f' to override

... This error repeats 39 times for all the other projects

```

And if you pass `-f` to force overwrite of course the last project always wins...

So IMHO the `toFile` / `--toFile=` should just be removed because I don't see a benefit in it anymore.

**Comment by mkurz** [comment]
Also, thinking about the `--config` flag: Would it make sense to name file `dependencies-scope.xx` (or `dependencies-Scope.xx`)? Or keep `dependencies.xx` for Compile but add the scope to the file name for any other scope?
Also would it be possible to make the flag not just take one scope, but multiple? so in one run it generates me multiple files?
Just thinking out loud:
```
> dependencyReport
# Writes target\dependencies.txt with the Compile scope
```

```
> dependencyReport --config=Test
# Writes target\dependencies-Test.txt with the Test scope (can be lower case, no preference from my side)
```

```
> dependencyReport --config=Compile,Test
# Writes BOTH target\dependencies.txt AND target\dependencies-Test.txt
```

Just and idea

**Comment by Friendseeker** [comment]
> Reason is because if you have multi-project builds, currently, with sbt 1.10.x, (when using toFile) it will just write the file for the first project and then complain for the next projects.

Thanks for the feedback! I think you are right and we shouldn't have `toFile`. It is not trivial to implement it properly (need to handle absolute/relative path, OS specific path, paths enclosed with single/double quote, tab completion...), and as you mentioned it completely breaks down for multi-project builds.

> Would it make sense to name file dependencies-scope.xx (or dependencies-Scope.xx)? Also would it be possible to make the flag not just take one scope, but multiple? so in one run it generates me multiple files?

These makes a lot of sense! Shall implement these 2 & make a draft PR to sbt repo for more feedbacks.

**Comment by mzuehlke** [comment]
Why are you diverging from the default way to select the scope ?
```Scala
> Compile / dependencyTree
> Test / dependencyTree
```

**Comment by Friendseeker** [comment]
> Why are you diverging from the default way to select the scope ?
> 
> ```scala
> > Compile / dependencyTree
> > Test / dependencyTree
> ```

Good point! Somehow this slipped my mind.

@mkurz What do you think? I personally think that sticking with the default is probably the better decision here so we don't introduce two distinct syntax for selecting scope, but that does mean we will have to part away from `--config=Compile,Test`. I plan to implement tab completion though so hopefully the downside of typing command twice is minimal.

Or maybe we can add a `--all` flag to select both `compile` and `test` scope (e.g. `dependencyReport --all`) and use the default syntax `Test / dependencyReport, Compile / dependencyReport` to select individual scope? Just brainstorming.

**Comment by mkurz** [comment]
Yeah, sure, we should just stick with the current way of selecting the scope. It seems I got a bit of tunnel vision while considering this issue...
Eventually, I think no flag is needed at all. _If_ you really want to have something like an "all" flag, then I would not use a flag for that but introduce an task called `dependencyReportAll` which runs the report for _all_ scopes that exists (not just Compile and Test)
This is the way it works with scalafmt and javafmt currently:
- https://scalameta.org/scalafmt/docs/installation.html#task-keys - `scalafmtAll`, `scalafmtCheckAll`
- https://github.com/sbt/sbt-java-formatter?tab=readme-ov-file#usage - `javafmtAll`, `javafmtCheckAll`

However, IMHO this isn't urgent right now and can be shipped later.

The file naming I suggest should still append the scope though to avoid collisions.

So in the end it seems there are two things we really want right now:
1. The `--format` flat to be able to select json or other formats.
2. Make sure an appropriate named file is written in the target folder (which currently does not happen).

**Comment by Friendseeker** [comment]
> Yeah, sure, we should just stick with the current way of selecting the scope. It seems I got a bit of tunnel vision while considering this issue... Eventually, I think no flag is needed at all. _If_ you really want to have something like an "all" flag, then I would not use a flag for that but introduce an task called `dependencyReportAll` which runs the report for _all_ scopes that exists (not just Compile and Test) This is the way it works with scalafmt and javafmt currently:
> 
> * https://scalameta.org/scalafmt/docs/installation.html#task-keys - `scalafmtAll`, `scalafmtCheckAll`
> * https://github.com/sbt/sbt-java-formatter?tab=readme-ov-file#usage - `javafmtAll`, `javafmtCheckAll`
> 
> However, IMHO this isn't urgent right now and can be shipped later.
> 
> The file naming I suggest should still append the scope though to avoid collisions.
> 
> So in the end it seems there are two things we really want right now:
> 
> 1. The `--format` flat to be able to select json or other formats.
> 2. Make sure an appropriate named file is written in the target folder (which currently does not happen).

Thanks. Shall immediately start drafting a PR based on your comment.
